[PlSql] Many updates to grammar

DmitryMikhailovich opened the pull request:
I've parsed more than 25k objects (primarily PL/SQL and DDL of views) from our enterprise Oracle DBs and discovered some missing features in this grammar.
This PR has the following kinds of changes:
- Support missing syntax rules and options.
- Support updated syntax (I've referenced Oracle 19 docs).
- Support undocumented syntax.
- Support obsolete (but still valid) syntax.
- Bugfixes.

It was tiresome to create separate PRs for each of the following 35 changes, so, please, forgive me for this huge PR🙂

- [cd394cb](https://github.com/antlr/grammars-v4/pull/3943/commits/cd394cb8bd0c07b4dedf5d760d8e386898184a4e) Support `EDITIONABLE`/`NONEDITIONABLE` options in `ALTER FUNCTION`.
- [536b367](https://github.com/antlr/grammars-v4/pull/3943/commits/536b3678908618147bb74abc1677d4750ae5c5ac) Support sharing clause in `CREATE FUNCTION`.
- [b9ea4d3](https://github.com/antlr/grammars-v4/pull/3943/commits/b9ea4d30fce783922be117a8118a309bb0d05d2d) Refactor [`PIPELINED USING` ](https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/PIPELINED-clause.html#GUID-FA182210-C68D-4E03-85B9-A6C681099705) and [`AGGREGATE USING`](https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/AGGREGATE-clause.html#GUID-2ED21240-E45A-4982-B674-CF0E1BE0985B) clauses in `CREATE FUNCTION` to match 19+ syntax.
- [9097ffc](https://github.com/antlr/grammars-v4/pull/3943/commits/9097ffc5589e5433f896c2c548f03d585e47d61e) Add optional exception handler in `CREATE PACKAGE BODY` initialize_section. [Documentation](https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/CREATE-PACKAGE-BODY-statement.html#GUID-68526FF2-96A1-4F14-A10B-4DD3E1CD80BE).
- [2e61913](https://github.com/antlr/grammars-v4/pull/3943/commits/2e619138da5e7d06999e304baa616e2e73a256a5) Add optional procedure properties in procedure definition spec. [Documentation](https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/procedure-declaration-and-definition.html#GUID-9A48D7CE-3720-46A4-B5CA-C2250CA86AF2).
- [5ae4c92](https://github.com/antlr/grammars-v4/pull/3943/commits/5ae4c92575d8ec187b760a4a66c9a939758fc2bc) Refactor options in function declaration spec (order of options does not matter); support `PARALLEL_ENABLE` clause.
- [f92f35f](https://github.com/antlr/grammars-v4/pull/3943/commits/f92f35fe5609ebcda6dafdc5619a393c58b15d51) Support [Conditional Compilation](https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/plsql-language-fundamentals.html#GUID-DC23291F-006D-49C4-87DA-6B31F19F162A). This adds only limited support for cases when conditional compilation directives are defined between statements, not in the middle of the statement.
- [30bb70f](https://github.com/antlr/grammars-v4/pull/3943/commits/30bb70fb682c3880de3423107424b5608d72996c) Add streaming clause in function body (that syntax is from 9i, but still valid. See examples in [old documentation](https://docs.oracle.com/cd/B10501_01/appdev.920/a96624/08_subs.htm#20554)).
- [126644e](https://github.com/antlr/grammars-v4/pull/3943/commits/126644ec21168d046153de81b13d21af884e36bf) Add optional `PARALLEL_ENABLE` clause in `CREATE PROCEDURE`. Though it is documented to be used only with procedures in packages, it is possible to create standalone procedure with `PARALLEL_ENABLE`. 
- [4874929](https://github.com/antlr/grammars-v4/pull/3943/commits/4874929e8cda8fafd17aaba5465e6e2f37993517) Fix compound form of `CREATE TRIGGER`: it is possible to have declarations in `timing_point_section` (fix of #3717). I've fixed the bug I've pulled myself🙈.
- [28bf697](https://github.com/antlr/grammars-v4/pull/3943/commits/28bf69775110613b9d5a2dddaed81db32a9ef699) Add undocumented `REFERENCES` alias for `REFERENCING` keyword in `CREATE TRIGGER`'s `referencing_clause`; support `referencing_element` to be optional if `REFERENCES` keyword is used. I haven't checked if it is possible to compile trigger with `REFERENCING` option without referencing element. By default if referencing elements are not specified, referencing names default to `NEW AS NEW OLD AS OLD` (as it can be seen in PLSQL Developer, in properties of the trigger).
![oracle_create_trigger_references](https://github.com/antlr/grammars-v4/assets/10853270/55f1b90c-d1e6-44ab-ad7d-e1328b548ec1)
- [b71ecb3](https://github.com/antlr/grammars-v4/pull/3943/commits/b71ecb3a9cbc497e81ff82443197394f2e495ffb) Support `TIMESTAMP` literals in `range_values_clause` and `index_partitioning_clause`.
- [e0b793f](https://github.com/antlr/grammars-v4/pull/3943/commits/e0b793fabe85ee6d43be68bc1f3182e6d7314121) Refactor `call_spec`, add 19+ syntax elements. [Documentation](https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/call-specification.html#GUID-C5F117AE-E9A2-499B-BA6A-35D072575BAD).
- [57c6fbb](https://github.com/antlr/grammars-v4/pull/3943/commits/57c6fbbf84223bbaf61d503ffab2064444627276) Support undocumented `NULL` option in `variable_declaration`.
- [78b16f0](https://github.com/antlr/grammars-v4/pull/3943/commits/78b16f00c53dc00b24955bda86c1b1cc37daa01d) Support optional `pragma_declaration` before `statement`. [Documentation](https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/plsql-language-fundamentals.html#GUID-D6EFD7E8-39DF-4430-B625-B6D37E49F6F4).
- [03e13fc](https://github.com/antlr/grammars-v4/pull/3943/commits/03e13fced8db822ce46a1ec20be9709a17578355) Use `general_element` instead of `collection_name` as an "argument" for `INDICES OF` clause.
- [f22d878](https://github.com/antlr/grammars-v4/pull/3943/commits/f22d87870bdcec3d5ae43271ff63337e0c6d06e0) Support [Collection Method Invocation](https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/collection-method.html#GUID-7AF1A3C4-D04B-4F91-9D7B-C92C75E3A300).
- [0bd64a5](https://github.com/antlr/grammars-v4/pull/3943/commits/0bd64a5f64b32c2bc38d5eafe1bfda615f23c170) Refactor `subquery_factoring_clause` into `with_clause` (19+ syntax). [Documentation](https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/SELECT.html#GUID-CFA006CA-6FF1-4972-821E-6996142A51C6). This could break visitors/listeners that depend on current syntax of `subquery_factoring_clause`.
- [f256d19](https://github.com/antlr/grammars-v4/pull/3943/commits/f256d19b5ec4dbe404f5c6087d3e29bd5cbc371f) [Support `GROUP BY` and hierarchical query clauses to be in any order](https://stackoverflow.com/questions/10043445/oracle-connect-by-clause-after-group-by-clause). 
- [becbc44](https://github.com/antlr/grammars-v4/pull/3943/commits/becbc44066c37bf0da60de67540c84da31d1734b) Support `LATERAL`.
- [0c6effc](https://github.com/antlr/grammars-v4/pull/3943/commits/0c6effc325587e3a40ba3292bb02adafca822334) Support [`CROSS/OUTER APPLY`](https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/SELECT.html#GUID-CFA006CA-6FF1-4972-821E-6996142A51C6__BABJHDDA) join.
- [6b53a72](https://github.com/antlr/grammars-v4/pull/3943/commits/6b53a724a7b89eb7e395652fcbe6d2320a1cf5de) Support `flashback_query_clause` syntax from 19+. [Documentation](https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/SELECT.html#GUID-CFA006CA-6FF1-4972-821E-6996142A51C6__I2126134).
- [63ccd30](https://github.com/antlr/grammars-v4/pull/3943/commits/63ccd3062ec6a492dcb526875db5127cb875bf55) Support [Multiset Operators](https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/Multiset-Operators.html#GUID-793FCBB0-A97C-4884-BCAC-DD0542EA746B).
- [3a282e3](https://github.com/antlr/grammars-v4/pull/3943/commits/3a282e34b9cb0aa9307e1865d959e25638fc17a7) Change leading `label_name` to `label_declaration` in `CASE` statements.
- [c7560ec](https://github.com/antlr/grammars-v4/pull/3943/commits/c7560ec421e0f5da4c8f6a16d172af6cf32df98d) Make precision optional for `NUMBER` type when it is used as `JSON_VALUE` return type. [Documentation](https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/JSON_VALUE.html#GUID-C7F19D36-1E75-4CB2-AE67-ADFBAD23CBC2).
- [0cf4bae](https://github.com/antlr/grammars-v4/pull/3943/commits/0cf4bae29a67b58c2c9cae5b7777b41533918bd8) Refactor `LAG/LEAD` analytic functions (support optional `respect_or_ignore_nulls`).
- [ecb3788](https://github.com/antlr/grammars-v4/pull/3943/commits/ecb3788b3450076f05a31618f54f23c063b7fd05) Support `NTH_VALUE`, `VAR_POP` and [undocumented `WM_CONCAT` ](https://oracle-base.com/articles/misc/string-aggregation-techniques#wm_concat) analytic functions.
- [f8d8e21](https://github.com/antlr/grammars-v4/pull/3943/commits/f8d8e215f84cd7e250f96fa9a897fad981722b07) Support more than one expression in `COLLECT` function's `ORDER BY` clause.
- [59013ab](https://github.com/antlr/grammars-v4/pull/3943/commits/59013abe3dfad16b23dab4e0054ba645b4779fdd) Change `expression` to a more generic `condition` rule in `where_clause`.
- [7692243](https://github.com/antlr/grammars-v4/pull/3943/commits/7692243accf8f8155925a04d46733734b42966c4) Refactor `general_element` to support [Object Access Expressions](https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/Object-Access-Expressions.html#GUID-FA69A056-12A6-420F-A106-EE252386CC43).
- [b1d2bef](https://github.com/antlr/grammars-v4/pull/3943/commits/b1d2bef38f74e372b67d8924407c7fb111fe32c5) Support more keywords as regular identifiers.
- [5c8390c](https://github.com/antlr/grammars-v4/pull/3943/commits/5c8390c31a2d354b51b0d6e30189ee20779fa2de) Add `%BULK_ROWCOUNT` lexer rule.
- [34de026](https://github.com/antlr/grammars-v4/pull/3943/commits/34de0260ebc8eaae954265a77627daa19fc3a9be) Support `~`, `/` and `\` as Q-string quote character.
- [b40f6fd](https://github.com/antlr/grammars-v4/pull/3943/commits/b40f6fdc4221b1ae7bedfc523daeefceccbd6b74) Support optional procedure properties in procedure body definition. [Documentation](https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/procedure-declaration-and-definition.html#GUID-9A48D7CE-3720-46A4-B5CA-C2250CA86AF2): `procedure_definition` Either defines a procedure that was declared earlier or both declares and defines a procedure.
- [8ee7fe8](https://github.com/antlr/grammars-v4/pull/3943/commits/8ee7fe85b4eb156b6a412c2150b6fc0432b69d3b) Support [Object Access expressions](https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/Object-Access-Expressions.html#GUID-FA69A056-12A6-420F-A106-EE252386CC43) on `TREAT` function's return value.

KvanTTT reviewed (commented):
> It was tiresome to create separate PRs for each of the following 35 changes, so, please, forgive me for this huge PR🙂

Thanks for the PR with such a clear and detailed history! I treat it an example for other contributors. But I see some commits are covered with tests some are not. Is it intentionally?

DmitryMikhailovich commented on sql/plsql/PlSqlLexer.g4:
```diff
@@ -2479,6 +2503,8 @@ START_CMD: // https://docs.oracle.com/cd/B19306_01/server.102/b14357/ch12002.htm
 
 REGULAR_ID: SIMPLE_LETTER (SIMPLE_LETTER | '$' | '_' | '#' | [0-9])*;
 
+INQUIRY_DIRECTIVE: '$$' (SIMPLE_LETTER | '_' )+;
```
It's a habit I've developed working on another grammar. 
Fixed.

DmitryMikhailovich commented on sql/plsql/PlSqlParser.g4:
```diff
@@ -492,7 +492,7 @@ alter_hierarchy
 
 alter_function
-    : ALTER FUNCTION function_name COMPILE DEBUG? compiler_parameters_clause* (REUSE SETTINGS)?
+    : ALTER FUNCTION function_name (EDITIONABLE | NONEDITIONABLE | COMPILE DEBUG? compiler_parameters_clause* (REUSE SETTINGS)? )
```
Fixed.

DmitryMikhailovich commented on sql/plsql/PlSqlParser.g4:
```diff
@@ -658,7 +668,7 @@ create_package_body
         IS
         | AS
-    ) package_obj_body* (BEGIN seq_of_statements)? END package_name?
+    ) package_obj_body* (BEGIN seq_of_statements (EXCEPTION exception_handler+)? )? END package_name?
```
Fixed.

DmitryMikhailovich commented on sql/plsql/PlSqlParser.g4:
```diff
@@ -1630,7 +1655,7 @@ global_partitioned_index
 
 index_partitioning_clause
-    : PARTITION partition_name? VALUES LESS THAN '(' literal (',' literal)* ')' segment_attributes_clause?
+    : PARTITION partition_name? VALUES LESS THAN '(' (literal (',' literal)* | TIMESTAMP literal (',' TIMESTAMP literal)* ) ')' segment_attributes_clause?
```
Fixed (list extracted into `index_partitioning_values_list` rule)

DmitryMikhailovich commented on sql/plsql/PlSqlParser.g4:
```diff
@@ -5349,25 +5374,45 @@ invoker_rights_clause
+    : (LANGUAGE C_LETTER | EXTERNAL)
+      ( NAME id_expression LIBRARY identifier
+      | LIBRARY identifier NAME id_expression
+      | LIBRARY identifier )
```
Fixed.

DmitryMikhailovich commented on sql/plsql/PlSqlParser.g4:
```diff
@@ -5582,6 +5628,24 @@ pipe_row_statement
+      | variable_declaration
+      | error_directive
+      | function_body
+      | procedure_body)+
```
Fixed.

DmitryMikhailovich commented on sql/plsql/PlSqlParser.g4:
```diff
@@ -5720,6 +5785,13 @@ savepoint_statement
+collection_method_call // collection methods invocation that could be used as a statement
+    : expression '.' DELETE ('(' index+=expression (',' index+=expression)* ')')?
+    | expression '.' EXTEND ('(' number=expression (',' index+=expression)* ')')?
+    | expression '.' TRIM ('(' index+=expression ')')?
```
Fixed.

DmitryMikhailovich commented on sql/plsql/PlSqlParser.g4:
```diff
@@ -5779,6 +5858,35 @@ cycle_clause
+
+filter_clause
+    : ( MEASURES
+      | hier_alias=object_name) TO condition
```
Fixed.

DmitryMikhailovich commented on sql/plsql/PlSqlParser.g4:
```diff
@@ -6321,6 +6432,8 @@ unary_expression
     | ALL unary_expression
     | /*TODO{(input.LA(1) == CASE || input.LA(2) == CASE)}?*/ case_statement /*[false]*/
+    | unary_expression '.' (COUNT | FIRST | LAST | LIMIT)
+    | unary_expression '.' (EXISTS | NEXT | PRIOR) '(' index+=expression ')'
```
Fixed.

DmitryMikhailovich commented on sql/plsql/PlSqlParser.g4:
```diff
@@ -5752,20 +5824,27 @@ explain_statement
+    : WITH (function_body | procedure_body)*
+      with_factoring_clause (',' with_factoring_clause)*
+    | WITH (function_body | procedure_body)+
+      (with_factoring_clause (',' with_factoring_clause)* )?
```
Fixed.

DmitryMikhailovich commented on sql/plsql/PlSqlParser.g4:
```diff
@@ -919,6 +943,7 @@ dml_event_nested_clause
 
 referencing_clause
     : REFERENCING referencing_element+
+    | REFERENCES referencing_element*
```
Fixed.
You mean to clarify usage of `REFERENCES` (add a comment into grammar code)?

DmitryMikhailovich commented:
> > It was tiresome to create separate PRs for each of the following 35 changes, so, please, forgive me for this huge PR🙂
> 
> Thanks for the PR with such a clear and detailed history! I treat it an example for other contributors.
> 
> But I see some commits are covered with tests some are not. Is it intentionally?

I've covered with tests (examples) the most non-trivial, obscure and undocumented cases. I can add examples for other changes, if it is necessary.

KvanTTT reviewed (commented):
> I can add examples for other changes, if it is necessary.

Ideally, we should have examples for all changes, and if it doesn't bother you, please add. Otherwise, I'll merge the PR without these tests, it's anyway very good.

KvanTTT commented:
@teverett please leave it to me, the PR should be merged using `Rebase and merge` method, it contains very clear commits.

DmitryMikhailovich commented:
> > I can add examples for other changes, if it is necessary.
> 
> Ideally, we should have examples for all changes, and if it doesn't bother you, please add. Otherwise, I'll merge the PR without these tests, it's anyway very good.

I'll add missing examples by the end of this week.

teverett commented:
@KvanTTT  leaving this on to you to merge :)

DmitryMikhailovich commented:
- [de1be77](https://github.com/antlr/grammars-v4/pull/3943/commits/de1be77987b77730f898495f02783565d6ea3c01) Fixed property list for package procedure definition and declaration (haven't read [docs ](https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/procedure-declaration-and-definition.html#GUID-9A48D7CE-3720-46A4-B5CA-C2250CA86AF2) attentively). 
![image](https://github.com/antlr/grammars-v4/assets/10853270/88237c40-1b7c-4726-911f-14789226d108)
- [973692d](https://github.com/antlr/grammars-v4/pull/3943/commits/973692d500854094dccfa1db5ead7d9231c921c6) Added optional `hierarchies_clause` after analytic view name in `FROM`.
- [1741015](https://github.com/antlr/grammars-v4/pull/3943/commits/17410158da678a77c965643c30ed3d2600267df8) Added rule for `SQL%BULK_ROWCOUNT` (in [5c8390c](https://github.com/antlr/grammars-v4/pull/3943/commits/5c8390c31a2d354b51b0d6e30189ee20779fa2de) I've added only lexer and forgot about parser).
- [bca4634](https://github.com/antlr/grammars-v4/pull/3943/commits/bca46345fc29e79bf34bc2587c49709481faa7b9) Added examples for all changes.

DmitryMikhailovich commented:
[60ef94a](https://github.com/antlr/grammars-v4/pull/3943/commits/60ef94affcb310b67aca596fbc7e189308ad8a6d) and [511143d](https://github.com/antlr/grammars-v4/pull/3943/commits/511143d25d14df9497a6ce9d1221ac508b2fbc7d)
Fixed regressions:
- added full syntax for `parallel_enable_clause` and `streaming_clause` to function declaration.
- `PIPELINED` and `DETERMINISTIC` options in function definition may be in any order with other function properties.
- removed `streaming_clause` from `create_function_body` which I've added in [30bb70f](https://github.com/antlr/grammars-v4/pull/3943/commits/30bb70fb682c3880de3423107424b5608d72996c) - confused create function rule with function spec rule.
Example sql/plsql/examples/create_package05.sql covers those changes and compiles as valid package (tested on live Oracle 19).

[8fc5378](https://github.com/antlr/grammars-v4/pull/3943/commits/8fc53788c6489fffa505f199e021ad005ea01011) Support any `expression` as a target of `IS [NOT] JSON`.

[aaa00b4](https://github.com/antlr/grammars-v4/pull/3943/commits/aaa00b4197499f4b2bc99365574e6b344aacda06) fixed typos and commented out some code in my examples (`sql_script` rule does not support cases, when there is other statement(s) after `/`

KvanTTT commented:
Thanks a lot!

Naros commented:
After these were applied to master, running `mvn test` in the `plsql` directory, the `alter_database.sql` example files with errors.  Are we certain that there are not any other regressions that may have been missed @DmitryMikhailovich  and @KvanTTT ?

KvanTTT commented:
It's quite strange because the current build on master is green.

Naros commented:
> It's quite strange because the current build on master is green.

I agree, I noticed the CI is green.  If you run `mvn test` from within the `plsql` directory manually, do you observe the same issue @KvanTTT ?
